### PR TITLE
2234 api for calculating detail daily statistic for userid

### PR DESCRIPTION
**Pull request by koylubaevnt**
#2234 BE: design and develop API for calculating detail daily statistic for userId

* Added endpoint "/v2/admin/study/day" for get user details daily statistic for user if you are admin AdminControllerV2.kt
* Added endpoint "/v2/statistics/study/day" for get current user details daily statistic UserStatisticControllerV2.kt
* Added native query StudyHistoryRepository.getDailyStatistics
* Added dto UserDailyDetailStatisticsDto for response
* Added projection UserDailyDetailStatisticsProjection to collect data from StudyHistoryRepository.getDailyStatistics method
* Added test for new features
* Added order by start time to get list of study history: last element is last date

### CODE REVIEW

File: src/main/kotlin/com/epam/brn/controller/AdminControllerV2.kt [hunk 1]
```diff
@@ -33,6 +33,16 @@ class AdminControllerV2(
     ) = ResponseEntity.ok()
         .body(BaseResponse(data = studyHistoryService.getHistories(userId, from, to)))
 
+    @GetMapping("/study/day")
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/controller/AdminControllerV2.kt:36]: could you please add into api.raml specification for this end-point

File: src/main/kotlin/com/epam/brn/dto/statistic/UserDailyDetailStatisticsDto.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.epam.brn.dto.statistic
+
+data class UserDailyDetailStatisticsDto(
+
+    val seriesName: String,
+
+    val doneExercises: Int,
+
+    val attempts: Int,
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/statistic/UserDailyDetailStatisticsDto.kt:9]: lets add description here into comments for this field

File: src/main/kotlin/com/epam/brn/model/UserDailyDetailStatisticsProjection.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    @get:Value("#{target.done_exercises}")
+    val doneExercises: Int?
+
+    val attempts: Int?
+
+    @get:Value("#{target.done_exercises_successfully_from_first_time}")
+    val doneExercisesSuccessfullyFromFirstTime: Int?
+
+    @get:Value("#{target.listen_words_count}")
+    val listenWordsCount: Int?
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/model/UserDailyDetailStatisticsProjection.kt:19]: can we make this fields not-nullable types?

File: src/main/kotlin/com/epam/brn/repo/StudyHistoryRepository.kt [hunk 4]
```diff
@@ -98,10 +99,35 @@ interface StudyHistoryRepository : CrudRepository<StudyHistory, Long> {
...
+                from (
+                    select sh.exercise_id, sh.tasks_count, sum(sh.replays_count) replays_count
+                    from study_history sh
+                    where sh.user_id = :userId and sh.start_time >= :from and sh.start_time <= :to
+                    group by sh.exercise_id, sh.tasks_count) t, exercise e, sub_group sg, series s
+                where t.exercise_id = e.id and e.sub_group_id = sg.id and sg.exercise_series_id = s.id
+                group by s.id, s.name """,
+        nativeQuery = true
+    )
+    fun getDailyStatistics(userId: Long, from: LocalDateTime, to: LocalDateTime): List<UserDailyDetailStatisticsProjection>
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/repo/StudyHistoryRepository.kt:128]: too complicate..
lets move calculation into Kotlin code by all studyHistory by day

File: src/main/kotlin/com/epam/brn/controller/AdminControllerV2.kt (hunk 1 above)
Comment (koylubaevnt) [reply src/main/kotlin/com/epam/brn/controller/AdminControllerV2.kt:36]: Done

File: src/main/kotlin/com/epam/brn/dto/statistic/UserDailyDetailStatisticsDto.kt (hunk 2 above)
Comment (koylubaevnt) [reply src/main/kotlin/com/epam/brn/dto/statistic/UserDailyDetailStatisticsDto.kt:9]: Done

File: src/main/kotlin/com/epam/brn/model/UserDailyDetailStatisticsProjection.kt (hunk 3 above)
Comment (koylubaevnt) [reply src/main/kotlin/com/epam/brn/model/UserDailyDetailStatisticsProjection.kt:19]: Done

File: src/main/kotlin/com/epam/brn/repo/StudyHistoryRepository.kt (hunk 4 above)
Comment (koylubaevnt) [reply src/main/kotlin/com/epam/brn/repo/StudyHistoryRepository.kt:128]: OK. I'll try

File: src/main/kotlin/com/epam/brn/model/UserDailyDetailStatisticsProjection.kt (hunk 3 above)
Comment (koylubaevnt) [reply src/main/kotlin/com/epam/brn/model/UserDailyDetailStatisticsProjection.kt:19]: If we change SQL (getDailyStatistics) to Kotlin code then don't need this class. I removed it

File: src/main/kotlin/com/epam/brn/repo/StudyHistoryRepository.kt (hunk 4 above)
Comment (koylubaevnt) [reply src/main/kotlin/com/epam/brn/repo/StudyHistoryRepository.kt]: Done

File: src/main/kotlin/com/epam/brn/service/StudyHistoryService.kt [hunk 5]
```diff
@@ -75,4 +92,52 @@ class StudyHistoryService(
...
     }
+
+    private fun calculateUserDailyDetailStatistics(studyHistories: List<StudyHistory>):
+        MutableList<UserDailyDetailStatisticsDto> {
+            val seriesIdToIndex = mutableMapOf<Long?, Int>()
+            val exerciseIdToIndex = mutableMapOf<Long?, Int>()
+            val repeatedExerciseIdToIndex = mutableSetOf<Long?>()
+            val result = mutableListOf<UserDailyDetailStatisticsDto>()
+            var index = 0
+            studyHistories.forEach {
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/service/StudyHistoryService.kt:103]: here you can make code more Kotlinish using

studyHistories.forEachIndexed { index, studyHistory -> ...}

?

File: src/main/kotlin/com/epam/brn/service/StudyHistoryService.kt [hunk 6]
```diff
@@ -75,4 +92,52 @@ class StudyHistoryService(
...
+                    val seriesId = it.exercise.subGroup!!.series.id
+                    val seriesIndex = seriesIdToIndex[seriesId]
+                    if (seriesIndex == null) {
+                        // New series -> create user statistic and fill it
+                        val userDailyDetailStatisticsDto = UserDailyDetailStatisticsDto(
+                            seriesName = it.exercise.subGroup!!.series.name,
+                            doneExercises = 1,
+                            attempts = it.replaysCount,
+                            doneExercisesSuccessfullyFromFirstTime = 1,
+                            listenWordsCount = it.tasksCount.toInt()
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/service/StudyHistoryService.kt:117]: listenWordsCount - it is actually count of resources which are linked to task.

File: src/main/kotlin/com/epam/brn/dto/statistic/UserDailyDetailStatisticsDto.kt [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package com.epam.brn.dto.statistic
+
+data class UserDailyDetailStatisticsDto(
+
+    /** Name of series */
+    var seriesName: String,
+
+    /** Count of done exercise. */
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/statistic/UserDailyDetailStatisticsDto.kt:8]: exercise -> exercises

File: src/main/kotlin/com/epam/brn/dto/statistic/UserDailyDetailStatisticsDto.kt [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    var doneExercises: Int,
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/statistic/UserDailyDetailStatisticsDto.kt:9]: doneExercises -> doneUniqueExercises

File: src/main/kotlin/com/epam/brn/dto/statistic/UserDailyDetailStatisticsDto.kt [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+    /** Count attempts to done exercise */
+    var attempts: Int,
```
Comment (ElenaSpb) [inline src/main/kotlin/com/epam/brn/dto/statistic/UserDailyDetailStatisticsDto.kt:12]: /** Repetiotion count of done exercises */
var repeatitionExercisesCount: Int,

File: src/main/kotlin/com/epam/brn/service/StudyHistoryService.kt (hunk 5 above)
Comment (koylubaevnt) [reply src/main/kotlin/com/epam/brn/service/StudyHistoryService.kt:103]: Yes, I can. But I don't need index here. What purpose to change this line of code?

Comment (ElenaSpb) [reply src/main/kotlin/com/epam/brn/service/StudyHistoryService.kt:103]: calculations looks too complicate,
maybe make it more kotlinish?

image

Comment (koylubaevnt) [reply src/main/kotlin/com/epam/brn/service/StudyHistoryService.kt]: OK. Made it. Please check

**Review by ElenaSpb — APPROVED** [review]

**Pull request merged**